pkg/virt-launcher/virtwrap/util/libvirt_helper.go

```diff
@@ -237,3 +251,124 @@ func NewDomain(dom cli.VirDomain) (*api.Domain, error) {
 	domain.GetObjectMeta().SetUID(domain.Spec.Metadata.KubeVirt.UID)
 	return domain, nil
 }
+
+// ForkAndMonitor itself to give qemu an extra grace period to properly terminate
+// in case of virt-launcher crashes
+func ForkAndMonitor(commandNamePrefix string) error {
+	cmd := exec.Command(os.Args[0], append(os.Args[1:], "--no-fork", "true")...)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+
+	done := make(chan error, 2)
+	go func() {
+		io.Copy(os.Stdout, stdout)
+		done <- err
+	}()
+
+	go func() {
+		io.Copy(os.Stderr, stderr)
+		done <- err
+	}()
+	if err := cmd.Start(); err != nil {
+		log.Log.Reason(err).Error("failed to fork virt-launcher")
+		return err
+	}
+
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT)
+	go func() {
+		for sig := range sigs {
+			log.Log.V(3).Log("signalling virt-launcher to shut down")
+			err := cmd.Process.Signal(syscall.SIGTERM)
+			if err != nil {
+				log.Log.Reason(err).Errorf("received signal %s but can't signal virt-launcher to shut down", sig.String())
+			}
+		}
+	}()
+
+	// stdout and stderr need to be fully read
+	<-done
+	<-done
+
+	if err := cmd.Wait(); err != nil {
+		exitCode := 1
+		if exiterr, ok := err.(*exec.ExitError); ok {
+			if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
+				exitCode = status.ExitStatus()
+			}
+		}
+		log.Log.Reason(err).Error("dirty virt-launcher shutdown")
+		pid, _ := virtlauncher.FindPid(commandNamePrefix)
```

Comment:
I think we want to check that the qemu process doesn't still exist regardless if virt-launcher exited cleanly or not.

Reply:
Good point. Will change it.